client/my-sites/add-ons/hooks/use-storage-add-ons.ts

```diff
@@ -0,0 +1,19 @@
+import { PRODUCT_1GB_SPACE } from '@automattic/calypso-products';
+import { useMemo } from '@wordpress/element';
+import useAddOns from './use-add-ons';
+
+interface Props {
+	siteId?: number | null;
+	isInSignup?: boolean;
+}
+
+const useStorageAddOns = ( { siteId, isInSignup }: Props ) => {
+	const addOns = useAddOns( siteId ?? undefined, isInSignup );
+
+	return useMemo(
+		() => addOns.filter( ( addOn ) => addOn?.productSlug === PRODUCT_1GB_SPACE ),
```

Comment:
I think the conditioning on `PRODUCT_1GB_SPACE` makes this a less generic function than what the title says. Should we maybe narrow it down a bit, introduce the slug as a parameter? 🤔 

@jeyip any thoughts about this? What is the connection between `PRODUCT_1GB_SPACE` and "storage add ons", and do you think we'd have more of these product slugs being relevant?

Reply:
> @jeyip any thoughts about this? What is the connection between PRODUCT_1GB_SPACE and "storage add ons", and do you think we'd have more of these product slugs being relevant?

Ah this is a fair question. To explain more about `PRODUCT_1GB_SPACE`, storage add-ons are represented as a quantity based item in checkout, so the 50GB storage add-on is equivalent to 

```javascript
// 50gb storage add-on
{
  product_slug: 'PRODUCT_1GB_SPACE',
  quantity: 50,
}

// 100gb storage add-on
{
  product_slug: 'PRODUCT_1GB_SPACE',
  quantity: 100,
}
```

[ To provide more context, this was an architectural decision made before I started on the project ]( https://github.com/Automattic/martech/issues/1907 ), and you'll note the same design throughout the add-ons data layer. I do understand how this could cause confusion though, especially without deeper knowledge of the storage add-ons implementation ( which isn't ideal ).

> I think the conditioning on PRODUCT_1GB_SPACE makes this a less generic function than what the title says. Should we maybe narrow it down a bit, introduce the slug as a parameter? 🤔

We _are_ mapping the 50GB and 100GB add-on to a more intuitive and human readable feature slug for the purposes of simpler information joining and tracking [here](https://github.com/Automattic/wp-calypso/blame/trunk/client/my-sites/add-ons/hooks/use-add-ons.ts#L134) and [here](https://github.com/Automattic/wp-calypso/blame/trunk/client/my-sites/add-ons/hooks/use-add-ons.ts#L148) ( no need to check product slug + quantity to identify individual storage add-ons ). Perhaps conditioning on featureSlugs would help in this situation?

If it would, I'm happy to make the necessary changes either here or in a quick follow-up.

```javascript
() =>
	addOns.filter( ( addOn ) => {
		const [ addOnSlug ] = addOn?.featureSlugs ?? [];
		return [ FEATURE_50GB_STORAGE_ADD_ON, FEATURE_100GB_STORAGE_ADD_ON ].includes( addOnSlug );
	} );
```

Reply:
Maybe we can just add the product slug as a parameter to  the `useAddOns` hook

Reply:
> This does seem to hint again at code smells with add-ons data layer design

Yeah, possibly. So it's great that we are looking into it.

> If it would, I'm happy to make the necessary changes either here or in a quick follow-up.

I think it's fine to ship this as-is for now. I'm personally not so concerned about the underneath for this hook i.e. whether it uses a product slug or feature slug for the filter. My concern looking at this, hence asking, is whether these slugs define some sort of "context" that will need to be elevated into a parameter of sorts. 

So if we expect to have different consumers of `useStorageAddOns` that will retrieve the add-ons differently (based on different slugs), then a couple of options for that:

1. Introduce "intent" (or "type" or something else) as a parameter, so `useStorageAddOns( { intent: 'jetpack-ai' } )`, etc. This feels a bit overkill if it's for 2 or less options.
2. Introduce a boolean "flag", so `useStorageAddOns( { isForJetpack: true } )`. Good for up to two options, but not for more.
3. Introduce separate hooks for each e.g. `useStorageAddOnsForJetpack`, etc.

but I'm not sure if that's the case (?) or if this will remain a single monolithic selection from the general add-ons hook (in which case it becomes a more low-level design concern how we filter these out) :-)

Reply:
Mostly curiosity! thanks for replies and explaining thoroughly :-)

Reply:
> My concern looking at this, hence asking, is whether these slugs define some sort of "context" that will need to be elevated into a parameter of sorts.

Ah I see what you're saying 👍 

> but I'm not sure if that's the case (?) or if this will remain a single monolithic selection from the general add-ons hook 

I'm currently uncertain either. The current roadmap indicates that they will remain a single monolithic selection for the near future, but things can always change 😅

Now that you've brought this up, I'll make an issue as a placeholder. As I continue work, I'll keep an eye out for different storage add-on upsell combinations. If we see more pop up, we can return to this discussion 🙂